### ISSUE TITLE
getRecords from PART of columns and random sort

### ISSUE THREAD

**Issue by paive-media** [thephpleague/csv#506]
## Bug Report

```php
use League\Csv\Reader;

$csv = <<<CSV
Abel,14,M,2004
Abiga,6,F,2004
Aboubacar,8,M,2004
Aboubakar,6,M,2004
CSV;

$reader = Reader::createFromString($csv);
$resultSet = Statement::create()->process($reader);
$records = $resultSet->getRecords([3 => 'Year', 0 => 'Firstname', 4 => 'Yolo']);
var_dump([...$records][0]);
//returns something like this
// array:4 [
//     "Year" => "2004",
//     "Firstname" => "Abel",
//     "Yolo" => null,
//  ]


// BUT 
var_dump([...$records][0]);
// returns
// array:4 [
//     "Year" => "Abel",
//     "Firstname" => "14",
//     "Yolo" => "M",
//  ]

```

| Information | Description |
|--------------|---------|
| Version | 0.9.11 |
| PHP version | 8.1 |
| OS Platform | CentOS 7 |

### Summary

Error in src/ResultSet.php function combineHeader
this one row is not enough for map fileds from `$header` condition
```php
$assocRecord = array_combine($header, $record);
```

I think this will be do the job:

```php
protected function combineHeader(array $header): Iterator
    {
        if ($header === $this->header || [] === $header) {
            return $this->records;
        }

        // not needed
        $field_count = count($header);
        $mapper = static function (array $record) use ($header, $field_count): array {

            // if (count($record) !== $field_count) {
            //     $record = array_slice(array_pad($record, $field_count, null), 0, $field_count);
            // }

            /** @var array<string, string|null> $assocRecord */
            // $assocRecord = array_combine($header, $record);
            
            // UPD 25-10-2023: get not linears PARTS for $header settings
            $assocRecord = [];
            // row data without assoc keys
            $tmp_record = array_values($record);
            foreach ($header as $col_idx => $new_record_key) {
                $assocRecord[$new_record_key] = $tmp_record[$col_idx]??null;

                // DEBUG
                // if(!isset($tmp_record[$col_idx])){
                //     var_dump($tmp_record);
                // }
            }

            return $assocRecord;
        };

        return new MapIterator($this->records, $mapper);
    }
```

**Comment by paive-media** [comment]
changes are also needed in src/Reader.php @ combineHeader

**Comment by nyamsprod** [comment]
@paive-media the bug you are reporting is not one.
The code you are expecting to get *will* only be available starting with version *9.12.0* which is not yet release.
